### PR TITLE
Adding learning rate schedulers to C++ API

### LINKED ISSUE #50577: Learning rate scheduler in C++ API
Hey!

Am I right in thinking that the learning rate scheduler is not available in the C++ API?

If not, are there any plans to implement it? Is there anything I can do to help with this?

Thanks 
James

cc @yf225 @glaringlee @vincentqb

### PR DESCRIPTION
**Pull request by jamesbut**
Fixes #50577

Learning rate schedulers had not yet been implemented for the C++ API. 

This pull request introduces the learning rate scheduler base class and the StepLR subclass. Furthermore, it modifies the existing OptimizerOptions such that the learning rate scheduler can modify the learning rate.

### CODE REVIEW

**Comment by glaringlee** [comment]
@jamesbut 
Thanks for contributing this PR, I'll start to review this.

There are some CI tests failures, please add all your new header files into here:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/optim.h
This should fix most of the failures.

**Comment by jamesbut** [comment]
@glaringlee 

I have already added the headers to the optim.h file. You should be able to see this in the changed files - without this the tests would not compile.

Do you have any other suggestions? It's difficult to debug because it compiles fine on my local machine.

**Comment by glaringlee** [comment]
> @glaringlee
> 
> I have already added the headers to the optim.h file. You should be able to see this in the changed files - without this the tests would not compile.
> 
> Do you have any other suggestions? It's difficult to debug because it compiles fine on my local machine.

I will figure it out. Update here later.

**Comment by glaringlee** [comment]
@jamesbut 
These failures are cpp extension build failures. You need to add the headers's path here as well.
https://github.com/pytorch/pytorch/blob/1aaddd83a501743aad18f2f36da2ba9a2c3eb317/setup.py#L920

**Review by glaringlee — COMMENTED** [review]
@jamesbut 
These failures are cpp extension build failures. You need to add the headers's path here as well.
https://github.com/pytorch/pytorch/blob/1aaddd83a501743aad18f2f36da2ba9a2c3eb317/setup.py#L920

**Comment by jamesbut** [comment]
@glaringlee 

Do you know how to solve these failures? They don't give errors like the previous failures

**Comment by glaringlee** [comment]
@jamesbut Don't worry about these failures, not related to your change. Sry about the delay, I got a bit busy recently on pytorch data related thing, I will try to finish my review by this Thursday.

File: torch/csrc/api/src/optim/schedulers/lr_scheduler.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+              "Number of learning rates not equal to the number of param groups\n",
+              "Number of learning rates given: ", learning_rates.size(),
+              "\nNumber of param groups: ", optimizer_.param_groups().size());
+
+  for(std::size_t i = 0; i < optimizer_.param_groups().size(); i++)
+    //optimizer_.param_groups()[i].options().lr() = learning_rates[i];
+    optimizer_.param_groups()[i].options().set_lr(learning_rates[i]);
+}
+
+const std::vector<double> LRScheduler::get_current_lrs() const {
```
Comment (glaringlee) [inline torch/csrc/api/src/optim/schedulers/lr_scheduler.cpp:29]: const std::vector<double> -> std::vector<double>

File: test/cpp/api/optim.cpp [hunk 2]
```diff
@@ -452,3 +452,55 @@ TEST(OptimTest, AddParameter_LBFGS) {
 
   // REQUIRE this doesn't throw
 }
+
+bool double_comparison(const double a, const double b,
```
Comment (glaringlee) [inline test/cpp/api/optim.cpp:456]: It seems this function is only being used by check_lr_change, can we move this inside the function?

**Review by glaringlee — COMMENTED** [review]
@jamesbut Generally looks good, some nit comments

**Comment by jamesbut** [comment]
Sorry for the late reply - I have been super busy recently. I have committed the changes that you suggested :)

**Comment by glaringlee** [comment]
@jamesbut Thank you so much for this contributing. I am going to approve this. Can you rebase your code with latest branch?

File: torch/csrc/api/include/torch/optim/schedulers/lr_scheduler.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  LRScheduler(torch::optim::Optimizer& optimizer);
+
+  void step();
+
+protected:
+  //A vector of learning rates is calculated and returned from the specific subclass.
+  //A vector is returned with each element being a separate learning rate for each
+  //param group - although the normal use case would be to return a vector of identical
+  //elements.
+  virtual std::vector<double> get_lrs() = 0;
```
Comment (glaringlee) [inline torch/csrc/api/include/torch/optim/schedulers/lr_scheduler.h:24]: Since you have this virtual function, you need to declare the destructor of this class as virtual, and the derived class (eg StepLR) will has virtual destructor automatically. Please change.

**Review by glaringlee — CHANGES_REQUESTED** [review]

**Review by glaringlee — APPROVED** [review]
LGTM now. Thanks for contributing this!

**Comment by jamesbut** [comment]
Great, thanks a lot for all your help :). Is there any further action that I have to take on my end?

**Comment by glaringlee** [comment]
> Great, thanks a lot for all your help :). Is there any further action that I have to take on my end?



> Great, thanks a lot for all your help :). Is there any further action that I have to take on my end?

Can you please rebase your code again? There was some issues in our master branch which made your code broke one of our CI test, a rebase can solve this problem. Thanks.

**Comment by glaringlee** [comment]
@jamesbut Can you please rebase this PR? There is a merge conflict in our internal system, need a rebase. Thanks a lot.

**Comment by jamesbut** [comment]
Aa awesome :), I will try and work on converting the other schedulers to C++ when I have a little more spare time

**Comment by glaringlee** [comment]
> Aa awesome :), I will try to work on converting the other schedulers to C++ when I have a little more spare time

Great! Thanks a lot for keep contributing to pytorch!
